posts/2023-09-07-generating-a-rust-client-library-for-zio-http-endpoints.md

```diff
@@ -0,0 +1,717 @@
+
+## ZIO Http
+
+If we rewrite `golem-cli` to Rust, we loose the convenience of using  **endpoint definitions** (written in Scala with ZIO Http, the ones we have for implementing the server) for calling our API, and we would also loose all the **types** used in these APIs as they are all defined as Scala case classes and enums. Just to have more context, let's take a look at one of the endpoints!
+
+A ZIO Http **endpoint** is just a definition of a single endpoint of a HTTP API, describing the routing as well the inputs and outputs of it:
+
+```scala
+val getWorkerMetadata =
+    Endpoint(GET / "v1" / "templates" / rawTemplateId / "workers" / workerName)
+      .header(Auth.tokenSecret)
+      .outErrorCodec(errorCodec)
+      .out[WorkerMetadata] ?? Doc.p("Get the current worker status and metadata")
+```
+
+Let's see what we have here:
+
+- the endpoint is reached by sending a **GET** request
+- the request **path** consists of some static segments as well as the _template id_ and the _worker name_ 
+- it also requires an **authorization header**
+- we define that kind of errors it can return with
+- and finally it defines that the responses **body** will contain a JSON representation (default in ZIO Http) of a type called `WorkerMetadata`
+
+What are `rawTemplateId` and `workerName`? These are so called **path codecs**, defined in a common place so they can be reused in multiple endpoints. They allow us to have dynamic parts of the request path mapped to specific types - so when we implement the endpoint (or call in a client) we don't have to pass strings, we can directly work with the business domain types, in this case `RawTemplateId` and `WorkerName`.
+
+The simplest way to define path codecs is to **transform** and existing one:
+
+```scala
+val workerName: PathCodec[WorkerName] =
+  string("worker-name").transformOrFailLeft(WorkerName.make(_).toErrorEither, _.value)
+```
+
+Here the `make` function is a **ZIO Prelude** [`Validation`](https://zio.github.io/zio-prelude/docs/functionaldatatypes/validation) which we have to convert to an `Either` for the transform function. Validations can contain more than one failures, opposed to `Either`s, which allows us to compose them in a way that we can keep multiple errors instead of immediately returning with the first failure.
```

Comment:
```suggestion
Here the `make` function is a **ZIO Prelude** [`Validation`](https://zio.github.io/zio-prelude/docs/functionaldatatypes/validation) which we have to convert to an `Either` for the transform function. Validations can contain more than one failures, as opposed to `Either`s, which allows us to compose them in a way that we can keep multiple errors instead of immediately returning with the first failure.
```